src/infill.cpp

```diff
@@ -263,45 +329,34 @@ void Infill::multiplyInfill(Polygons& result_polygons, Polygons& result_lines)
 
 void Infill::generateGyroidInfill(Polygons& result_lines)
 {
-    GyroidInfill::generateTotalGyroidInfill(result_lines, zig_zaggify, outline_offset + infill_overlap, infill_line_width, line_distance, in_outline, z);
+    GyroidInfill::generateTotalGyroidInfill(result_lines, zig_zaggify, line_distance, inner_contour, z);
 }
 
-void Infill::generateConcentricInfill(Polygons& result, int inset_value)
+void Infill::generateConcentricInfill(VariableWidthPaths& toolpaths, const Settings& settings)
 {
-    Polygons first_concentric_wall = in_outline.offset(outline_offset + infill_overlap - line_distance + infill_line_width / 2); // - infill_line_width / 2 cause generateConcentricInfill expects [outline] to be the outer most polygon instead of the outer outline
-
-    if (perimeter_gaps)
+    constexpr coord_t wall_0_inset = 0; // Don't apply any outer wall inset for these. That's just for the outer wall.
+    const bool iterative = line_distance > infill_line_width; // Do it all at once if there is not need for a gap, otherwise, iterate.
+    const coord_t min_area = infill_line_width * infill_line_width;
+    Polygons current_inset = inner_contour.offset(infill_line_width / 2);
+    do
     {
-        const Polygons inner = first_concentric_wall.offset(infill_line_width / 2 + perimeter_gaps_extra_offset);
-        const Polygons gaps_here = in_outline.difference(inner);
-        perimeter_gaps->add(gaps_here);
-    }
-    generateConcentricInfill(first_concentric_wall, result, inset_value);
-}
-
-void Infill::generateConcentricInfill(Polygons& first_concentric_wall, Polygons& result, int inset_value)
-{
-    result.add(first_concentric_wall);
-    Polygons* prev_inset = &first_concentric_wall;
-    Polygons next_inset;
-    Polygons new_inset;  // This intermediate inset variable is needed because prev_inset is referencing
-    while (prev_inset->size() > 0)
-    {
-        new_inset = prev_inset->offset(-inset_value);
-        new_inset.simplify();
-        result.add(new_inset);
-        if (perimeter_gaps)
+        if (iterative)
         {
-            const Polygons outer = prev_inset->offset(-infill_line_width / 2 - perimeter_gaps_extra_offset);
-            const Polygons inner = new_inset.offset(infill_line_width / 2);
-            const Polygons gaps_here = outer.difference(inner);
-            perimeter_gaps->add(gaps_here);
+            current_inset = current_inset.offset(-infill_line_width * 2).offset(infill_line_width * 2);
         }
-        // This operation helps to prevent the variable "prev_inset" changes whenever next_inset changes
-        next_inset = new_inset;
-        prev_inset = &next_inset;
-    }
-    std::reverse(std::begin(result), std::end(result));
+        current_inset.simplify();
+        if (current_inset.area() <= min_area)
+        {
+            break;
+        }
+
+        const coord_t inset_wall_count = iterative ? 1 : std::numeric_limits<coord_t>::max();
```

Comment:
When `iterative == true`, the beady strategy factory gets `max_bead_count = -2` in [WallToolPaths::generate()](https://github.com/Ultimaker/CuraEngine/pull/1210/files#diff-7f80a096e5681f5a38e9bf0844c79db22dc0580d2ecb809505ff34a595fefb9cR78). This is because:
```c++
 coord_t max_bead_count = 2 * inset_count = 2 * std::numeric_limits<coord_t>::max() = -2
 ```

I'm unsure about the consequences. This ends up skipping the `RedistributeBeadingStrategy` and the `LimitedBeadingStrategy` wrappers for concentric infills, as they require `max_bead_count > 0`. The intent told by the code would be that `max_bead_count` is saturating at the maximum value, but I don't think that makes sense either.

`-ftrapv` and `-fsanitize=signed-integer-overflow` won't catch this overflow since it happens on a size_t:  coord_t -> size_t (overflow) -> coord_t :wink:

Reply:
If the intent was indeed `max_bead_count >> 0`, it slightly improves the concentric surfaces:

![beforeafter](https://user-images.githubusercontent.com/3748582/130465793-6a3e4eba-ad73-4aed-abba-df582069b6a3.png)

Right slice has `RedistributeBeadingStrategy` and `LimitedBeadingStrategy` applied over inward distributed strategy. 

<details>
<summary>rough patch</summary>

```diff
diff --git a/src/BeadingStrategy/BeadingStrategyFactory.cpp b/src/BeadingStrategy/BeadingStrategyFactory.cpp
index 13f97b8cd..9c97317ee 100644
--- a/src/BeadingStrategy/BeadingStrategyFactory.cpp
+++ b/src/BeadingStrategy/BeadingStrategyFactory.cpp
@@ -17,6 +17,9 @@ namespace cura
 
 coord_t getWeightedAverage(const coord_t preferred_bead_width_outer, const coord_t preferred_bead_width_inner, const coord_t max_bead_count)
 {
+    if(preferred_bead_width_outer == preferred_bead_width_inner)
+    {
+        return preferred_bead_width_outer;
+    }
     if (max_bead_count > 2)
     {
         return ((preferred_bead_width_outer * 2) + preferred_bead_width_inner * (max_bead_count - 2)) / max_bead_count;
diff --git a/src/infill.cpp b/src/infill.cpp
index ccd62c972..9c1d44fe6 100644
--- a/src/infill.cpp
+++ b/src/infill.cpp
@@ -350,7 +350,7 @@ void Infill::generateConcentricInfill(VariableWidthPaths& toolpaths, const Setti
             break;
         }
 
-        const coord_t inset_wall_count = iterative ? 1 : std::numeric_limits<coord_t>::max();
+        const coord_t inset_wall_count = iterative ? 1 : std::numeric_limits<coord_t>::max() / 2;
         WallToolPaths wall_toolpaths(current_inset, infill_line_width, inset_wall_count, wall_0_inset, settings);
         const VariableWidthPaths inset_paths = wall_toolpaths.getToolPaths();
```
</details>

Reply:
Good find!

Reply:
This is why we should be using unsigned integers for counts of stuff, like number of beads :(

Reply:
This is fixed in a slightly different way than the suggestion here: https://github.com/Ultimaker/CuraEngine/pull/1210/commits/549a7f5446a51443d0516dab6b029edd6479b5ce

* Instead of limiting the wall count in `generateConcentricInfill`, I've limited it in `WallToolPaths::generate`. That way we don't need to keep this limit in mind when calling that function.
* The fix in the weighted average calculation is a bit more generic now, such that it would also still work if the outer and inner line widths are different, but the maximum bead count still so big it would cause integer overflows in the weighted sum.

Reply:
I've just realized that a warning is triggered because `max_bead_count = std::numeric_limits<coord_t>::max()` is odd. Although, this shouldn't be a problem since the bead count will not reach that value.

Reply:
Yeah it's using `coord_t` (the type we use for coordinates in 2D) with a count of stuff, for which we'd usually use `size_t`. `size_t` is unsigned.

It's never really going to hit the limit of 9 billion billion lines of course, but the advantage of using correct types is mainly to communicate to the developer reading this code what the value is, and to allow the compiler to warn us when we're mixing things up.